### PR TITLE
feat(recurring-job): set recurring job to pvc

### PR DESCRIPTION
**Pull request by c3y1huang**
https://github.com/longhorn/longhorn/discussions/5791

### CODE REVIEW

**Comment by c3y1huang** [comment]
Test result: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/3860/

**Comment by c3y1huang** [comment]
[Test PVC recurring job sync from Volume after upgrade](https://github.com/longhorn/longhorn-tests/blob/2d0aa0d45e0c2ea3810f2f3e95e5df0d65361f03/docs/content/manual/release-specific/v1.5.0/test-pvc-recurring-job-sync-after-upgrade.md): PASS

**Review by innobead — COMMENTED** [review]
Can we make the implementation simpler as below w/ fewer resources (w/o a new PVC controller), WDYT?

1. in volume_controller, when resync, check if the volume has PVC and it has recurring jobs labels
2. If yes, override the volume labels, because PVC labels have higher priority.
3. If no PVC, all things are the same as usual

Just need to define the priority for which resource will decide the labels, no need to do sync efforts which will require another PVC controller in the implementation and seem not really necessary cost. Still, we can achieve the case for gitops use case.

**Comment by c3y1huang** [comment]
Test result: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/3883/

File: datastore/longhorn.go [hunk 1]
```diff
@@ -724,27 +724,28 @@ func MarshalLabelToVolumeRecurringJob(labels map[string]string) map[string]*long
...
+		}
+		logrus.Debugf("Added volume %v recurring job label %v", volume.Name, labelKey)
 	}
-	if _, exist := v.Labels[key]; exist {
-		delete(v.Labels, key)
-		v, err = s.UpdateVolume(v)
+	return nil
+}
+
+func (s *DataStore) DeleteRecurringJobLabelOn(volume *longhorn.Volume, labelKey string) (*longhorn.Volume, error) {
```
Comment (innobead) [inline datastore/longhorn.go:739]: `DeleteRecurringJobLabelFromVolume`

File: types/types.go [hunk 2]
```diff
@@ -500,6 +500,13 @@ func GetVolumeLabels(volumeName string) map[string]string {
 	}
 }
 
+func GetRecurringJobLabelKeyBy(name string, isGroup bool) string {
```
Comment (innobead) [inline types/types.go:504]: `GetRecurringJobLabelKeyByType`

File: controller/recurring_job_controller.go [hunk 3]
```diff
@@ -497,3 +500,52 @@ func (control *RecurringJobController) newCronJob(recurringJob *longhorn.Recurri
 
 	return cronJob, nil
 }
+
+func syncRecurringJobLabelsTo(targetKind string, targetObj, sourceObj runtime.Object, log logrus.FieldLogger) error {
```
Comment (innobead) [inline controller/recurring_job_controller.go:538]: `syncRecurringJobLabelsFromSourceResource`

File: controller/volume_controller.go [hunk 4]
```diff
@@ -3455,22 +3455,49 @@ func (vc *VolumeController) updateRecurringJobs(v *longhorn.Volume) (err error)
...
 	if err == nil && !reflect.DeepEqual(existingVolume.Labels, v.Labels) {
 		_, err = vc.ds.UpdateVolume(v)
 	}
 
 	return nil
 }
 
+func (vc *VolumeController) syncPVCRecurringJobLabels(volume *longhorn.Volume) error {
+	kubeStatus := volume.Status.KubernetesStatus
+	if kubeStatus.PVCName == "" {
```
Comment (innobead) [inline controller/volume_controller.go:3483]: Do we need to check if the PVC still bound to the volume via `LastPVCRefAt`?

File: types/types.go [hunk 5]
```diff
@@ -511,6 +518,12 @@ func GetRecurringJobLabelValueMap(labelType, recurringJobName string) map[string
 	}
 }
 
+func IsRecurringJobLabel(key string) bool {
```
Comment (innobead) [inline types/types.go:528]: Add doc.

File: manager/volume.go [hunk 6]
```diff
@@ -717,11 +708,21 @@ func (m *VolumeManager) ListVolumeRecurringJob(volumeName string) (map[string]*l
...
 		return nil, err
 	}
-	return m.ds.DeleteVolumeRecurringJob(name, isGroup, volume)
+
+	key := types.GetRecurringJobLabelKeyBy(name, isGroup)
+	volume, err = m.ds.DeleteRecurringJobLabelOn(volume, key)
+	if err != nil {
+		return nil, err
+	}
+	return m.ds.UpdateVolume(volume)
```
Comment (innobead) [inline manager/volume.go:725]: Why not just updating the volume in datastore's `DeleteRecurringJobLabelOn`?

File: controller/recurring_job_controller.go [hunk 7]
```diff
@@ -275,20 +261,71 @@ func (control *RecurringJobController) cleanupVolumeRecurringJob(recurringJob *l
...
+func (control *RecurringJobController) getIsLastGroupsExistsIn(recurringJob *longhorn.RecurringJob, existingRecurringJobs map[string]*longhorn.RecurringJob) map[string]bool {
+	groups := make(map[string]bool, len(recurringJob.Spec.Groups))
+	for _, group := range recurringJob.Spec.Groups {
+		groups[group] = true
+		for _, existingRecurringJob := range existingRecurringJobs {
+			if existingRecurringJob.Name == recurringJob.Name {
+				continue
+			}
+			if util.Contains(existingRecurringJob.Spec.Groups, group) {
+				groups[group] = false
```
Comment (innobead) [inline controller/recurring_job_controller.go:305]: break?

**Comment by c3y1huang** [comment]
Test result: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/3911/

File: controller/recurring_job_controller.go [hunk 8]
```diff
@@ -226,116 +229,171 @@ func (control *RecurringJobController) syncRecurringJob(key string) (err error)
...
 		log.WithError(err).Error("failed to reconcile recurring job")
 	}
 
 	return nil
 }
 
-func (control *RecurringJobController) cleanupVolumeRecurringJob(recurringJob *longhorn.RecurringJob) error {
+func (c *RecurringJobController) cleanupRecurringJobLabelFromVolumeAndPVC(recurringJob *longhorn.RecurringJob) (err error) {
+	defer func() {
+		err = errors.Wrap(err, "failed to cleanup recurring job labels from Volumes and PVCs")
```
Comment (innobead) [inline controller/recurring_job_controller.go:249]: `"failed to cleanup recurring job labels %v from Volumes and PVCs"`

File: controller/recurring_job_controller.go [hunk 9]
```diff
@@ -226,116 +229,171 @@ func (control *RecurringJobController) syncRecurringJob(key string) (err error)
...
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err = errors.Wrapf(err, "failed to update recurring job labels from Volume %v to PVC %v", volume.Name, pvc.Name)
+	}()
+
+	if err := syncRecurringJobLabelsToTargetResource(types.KubernetesKindPersistentVolumeClaim, pvc, volume, c.logger); err != nil {
+		return errors.Wrap(err, "failed to sync recurring job labels to PVC")
```
Comment (innobead) [inline controller/recurring_job_controller.go:345]: `"failed to sync recurring job labels to PVC %v"`

File: controller/volume_controller.go [hunk 10]
```diff
@@ -3448,51 +3450,88 @@ func (vc *VolumeController) shouldRestoreRecurringJobs(v *longhorn.Volume) bool
...
-func (vc *VolumeController) updateRecurringJobs(v *longhorn.Volume) (err error) {
+func (c *VolumeController) updateRecurringJobs(v *longhorn.Volume) (err error) {
 	defer func() {
 		err = errors.Wrapf(err, "failed to update recurring jobs for %v", v.Name)
 	}()
 
 	existingVolume := v.DeepCopy()
+
+	if err := c.syncPVCRecurringJobLabels(v); err != nil {
+		return errors.Wrap(err, "failed attempting to sync PVC recurring job labels")
```
Comment (innobead) [inline controller/volume_controller.go:3474]: `failed attempting to sync PVC recurring job labels from volume %v`

File: controller/recurring_job_controller.go [hunk 11]
```diff
@@ -226,116 +229,171 @@ func (control *RecurringJobController) syncRecurringJob(key string) (err error)
...
+			if util.Contains(existingRecurringJob.Spec.Groups, group) {
+				unusedGroupInfo[group] = false
+				break
 			}
 		}
 	}
+	return unusedGroupInfo
+}
+
+func (c *RecurringJobController) updatePVCRecurringJobLabelsFrom(volume *longhorn.Volume) (err error) {
```
Comment (innobead) [inline controller/recurring_job_controller.go:329]: Rename `updatePVCRecurringJobLabelsFromVolume`

**Review by innobead — CHANGES_REQUESTED** [review]
Some comments need to resolve.

I forgot why we couldn't remove a group label from a volume if the group is being used. What does that matter?

**Comment by c3y1huang** [comment]
> I forgot why we couldn't remove a group label from a volume if the group is being used. What does that matter?

When a RecurringJob CR gets deleted, we also cleans the labels in the Volumes label. So we need to ensure the group label has no-op (no recurring jobs).

**Comment by innobead** [comment]
> > I forgot why we couldn't remove a group label from a volume if the group is being used. What does that matter?
> 
> When a RecurringJob CR gets deleted, we also cleans the labels in the Volumes label. So we need to ensure the group label has no-op (no recurring jobs).

I see, so it's about removing the unused group label from other volumes.

**Review by innobead — APPROVED** [review]

**Pull request merged**